bloom-housing/bloom
# 2250/auth cookie

# Pull Request Template

## Issue Overview

This PR addresses #2250


## Description

Refactors authentication flow from session storage to a secured cookie. The reason why we should use the cookie, instead of other options is described here: https://tkacz.pro/how-to-securely-store-jwt-tokens/

I've refactored backend & frontend apps. Once a user tries to authenticate, a backend creates a cookie with httpOnly and secure options and sends along with a response. The cookie is visible in browser cookies tab, but there is no way to read the cookie using javascript.

From the frontend perspective, to authorize requests I had to add the `withCredentials` parameter to an axios config. It sends cookies along with a request.

To log out a user, I defined a new `/logout` endpoint that creates a new cookie with the current date and sends in response. Because a cookie date is set to the current date, it means it's expired to a user is logged out.

## How Can This Be Tested/Reviewed?
The easiest way to test this is to go through a few of the different log in processes

try to log into the public site (as a non-mfa user) then try opening a new tab. You should remain logged in to both without needing to relog in

try to log into the partner site (as an mfa user and later as a nonmfa user). Try opening a new tab, and you should see that user remains logged in the new tab. 

you should also try to make sure that if the user is a jurisdictional admin then the permissions should still work correctly 

## Checklist:


## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue, and ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.
